File: Python/data_structures/09_Rabin_Karp.py
import sys
from random import randint, seed

PRIME = 1000000007


def polynomial_hashing(string, prime, x):
    value = 0
    for letter in reversed(string):
        value = (value * x + ord(letter)) % prime
    return value


def calc_power_of_x(x, prime, exponent):
    power = 1
    for _ in range(exponent):
        power = (power * x) % prime
    return power


def precompute_hashes_for_substrings_of_text(text, text_length, pattern_length, p, x):
    hash_values = [0] * (text_length - pattern_length + 1)
    last_substring_in_text = text[-pattern_length:]
    hash_values[-1] = polynomial_hashing(last_substring_in_text, p, x)
    x_power = calc_power_of_x(x, p, pattern_length)
    for ix in range(text_length-pattern_length-1, -1, -1):
        hash_values[ix] = (x*hash_values[ix+1] + ord(text[ix]) - x_power*ord(text[ix+pattern_length])) % p
    return hash_values


def rabin_karp_algorithm(text, pattern, prime):
    x = randint(1, prime-1)
    pattern_length = len(pattern)
    text_length = len(text)
    matching_indices = []
    pattern_hash = polynomial_hashing(pattern, prime, x)
    substring_hashes = precompute_hashes_for_substrings_of_text(text, text_length, pattern_length, prime, x)
    for ix in range(text_length-pattern_length+1):
        if pattern_hash == substring_hashes[ix] and text[ix:ix+pattern_length] == pattern:
            matching_indices.append(ix)
    return matching_indices


def main():
    data = sys.stdin.read().splitlines()
    pattern = data[0]
    text = data[1]
    seed(2112)
    result = rabin_karp_algorithm(text, pattern, PRIME)
    print(" ".join([str(ix) for ix in result]))


if __name__ == "__main__":
    main()

File: Rust/data_structures/src/bin/09_Rabin_Karp.rs
use std::io;
use rand::{RngExt, SeedableRng};
use rand::rngs::StdRng;

fn read_line() -> String {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();
    line.trim().to_string()
}

fn polynomial_hashing(string: &str, prime: i64, x: i64) -> i64 {
    string
        .bytes()
        .rev()
        .fold(0, |acc, code| (acc * x + code as i64) % prime)
}

fn calc_power_of_x(x: i64, exponent: usize, prime: i64) -> i64 {
    (0..exponent).fold(1, |acc, _| acc * x % prime)
}

fn calc_substring_hashes(text: &str, text_length: usize, pattern_length: usize, prime: i64, x: i64) -> Vec<i64> {
    let length = text_length - pattern_length + 1;
    let mut hash_values = vec![0; length];
    let last_substring_in_text = &text[(text_length-pattern_length)..text_length];
    hash_values[length - 1] = polynomial_hashing(last_substring_in_text, prime, x);
    let x_power = calc_power_of_x(x, pattern_length, prime);
    let ascii = text.as_bytes();
    for ix in (0..(length - 1)).rev() {
        let value = x * hash_values[ix+1] + ascii[ix] as i64 - x_power * ascii[ix + pattern_length] as i64;
        hash_values[ix] = value.rem_euclid(prime);
    };

    hash_values
}

fn rabin_karp_algorithm(text: &str, pattern: &str, prime: i64) -> Vec<usize> {
    let mut rng = StdRng::seed_from_u64(2112);
    let x = rng.random_range(1..prime);
    let pattern_length = pattern.len();
    let text_length = text.len();
    let pattern_hash = polynomial_hashing(pattern, prime, x);
    let substring_hashes = calc_substring_hashes(text, text_length, pattern_length, prime, x);
    let mut matching_indices = Vec::new();
    for ix in 0..(text_length-pattern_length+1) {
        if pattern_hash == substring_hashes[ix] && &text[ix..(ix+pattern_length)] == pattern {
            matching_indices.push(ix);
        }
    }

    matching_indices
}

fn main() {
    let pattern = read_line();
    let text = read_line();
    let prime: i64 = 1000000007;
    let result = rabin_karp_algorithm(&text, &pattern, prime);
    println!("{}", result.iter().map(|x| x.to_string()).collect::<Vec<_>>().join(" "));
}
